src/Analyzer/Passes/QueryAnalysisPass.cpp

```diff
@@ -4717,13 +4715,29 @@ ProjectionNames QueryAnalyzer::resolveExpressionNode(QueryTreeNodePtr & node, Id
             {
                 node = tryResolveIdentifier({unresolved_identifier, IdentifierLookupContext::TABLE_EXPRESSION}, scope).resolved_identifier;
 
-                /// If table identifier is resolved as CTE clone it
-                bool resolved_as_cte = node && node->as<QueryNode>() && node->as<QueryNode>()->isCTE();
+                /// If table identifier is resolved as CTE clone it and resolve
+                auto * subquery_node = node->as<QueryNode>();
+                auto * union_node = node->as<UnionNode>();
+                bool resolved_as_cte = (subquery_node && subquery_node->isCTE()) || (union_node && union_node->isCTE());
 
                 if (resolved_as_cte)
                 {
                     node = node->clone();
-                    node->as<QueryNode &>().setIsCTE(false);
+                    subquery_node = node->as<QueryNode>();
+                    union_node = node->as<UnionNode>();
+
+                    if (subquery_node)
+                        subquery_node->setIsCTE(false);
+                    else
+                        union_node->setIsCTE(false);
+
+                    IdentifierResolveScope subquery_scope(node, &scope /*parent_scope*/);
+                    subquery_scope.subquery_depth = scope.subquery_depth + 1;
+
+                    if (subquery_node)
+                        resolveQuery(node, subquery_scope);
+                    else
+                        resolveUnion(node, subquery_scope);
```

Comment:
In the above block, we check multiple times whether node is QueryNode or UnionNode, mostly because of CTE manipulation.I agree this is better than duplicating code by having a big if else block.
If it's likely to be used extensively like this, maybe we can have an abstract parent to QueryNode & UnionNode which supports CTE ( in a separate PR), what do you think ?

Reply:
We discussed it with @novikd before, and I am not sure about this. Query and Union nodes have some common boolean flags, so this can be performed. But need to discuss once again.